lib/tokenize.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+const split = String.prototype.split;
+const numbers = new Set(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']);
+
+const DEFAULT = 0;
+const IN_SQUARE_BRACKETS = 1;
+const IMMEDIATELY_AFTER_SQUARE_BRACKETS = 2;
+let cache = undefined;
+
+function tokenize(str) {
+  if (str.indexOf('[') === -1) {
+    if (str.indexOf('.') !== -1) {
+      return split.call(str, '.');
+    }
+    return [str];
+  }
+
+  const result = [],
+      len = str.length;
+
+  let curPropertyName = '',
+      i = 0,
+      state = 0,
+      char = '';
+
+  for (i = 0; i < len; ++i) {
+    char = str[i];
+
+    // Fall back to treating as property name rather than bracket notation if
+    // square brackets contains something other than a number.
+    if (state === IN_SQUARE_BRACKETS && char !== ']' && !numbers.has(char)) {
+      state = DEFAULT;
+      curPropertyName = result[result.length - 1] + '[' + curPropertyName;
+      result.pop();
+    }
+
+    if (char === '[') {
+      if (state !== IMMEDIATELY_AFTER_SQUARE_BRACKETS) {
+        result.push(curPropertyName);
+        curPropertyName = '';
+      }
+      state = IN_SQUARE_BRACKETS;
+    } else if (char === ']') {
+      if (state === IN_SQUARE_BRACKETS) {
+        state = IMMEDIATELY_AFTER_SQUARE_BRACKETS;
+        result.push(curPropertyName);
+        curPropertyName = '';
+      } else {
+        state = DEFAULT;
+        curPropertyName += char;
+      }
+    } else if (char === '.') {
+      if (state !== IMMEDIATELY_AFTER_SQUARE_BRACKETS) {
+        result.push(curPropertyName);
+        curPropertyName = '';
+      }
+      state = DEFAULT;
+    } else {
+      curPropertyName += char;
+    }
+  }
+
+  if (state !== IMMEDIATELY_AFTER_SQUARE_BRACKETS) {
+    result.push(curPropertyName);
+  }
+
+  return result;
+}
+
+function tokenizeCached(str) {
+  const result = cache.get(str);
+  if (result) {
+    return result;
+  }
+  cache.set(str, tokenize(str));
+  return cache.get(str);
+}
+
+exports.tokenize = tokenize;
+
+exports.setTokenizeCache = function(parmCache) {
```

Comment:
Avoid this and use an LRU-cache from start, the one from mnemonist is great.

Reply:
@mcollina Yes! mnemonist is truelly faster.

Should I use the new Function Syntax only for the tokenize method? I use some one time initialized sets and the cache, but it seems that I have to somehow pass them into the method? Do you have some example, where i could look how to it properly?
Thx.

Reply:
Check out the fast-json-stringify module. I'd cache the full thing.

Reply:
@mcollina 
I unfortunately don't know how I could apply your suggested change. I know you are a busy man, but could you help me out by giving me the code solution please? I am eager to learn as I want to improve more packages over the time, so it would be not a waste of your time, imho.

Please.